File: Orange/W4_DevideAndConquer/7_Distance in Tree.py
# f[u][dist] = f_s[u][dist] + (f[parent[u]][dist-1]-f_s[u][dist-2])
# f_s[u][0]=1
# f[root][0]=1
# f[i][0]=1
# f[root][dist]=f_sub[root][dist] for i in range(k+1)
# ans = tong f[i][dist] for i in range(0,N):
# ans/2 chinh la dap an
# tham khao goi y:https://codeforces.com/blog/entry/76080?#comment-604916
# sol nay dc AC neu viet bang C++, tren python se bi TLE hoac MLE
import sys

sys.setrecursionlimit(500000)
topo = []


def cal_f_sub(root, p, parent, Adj, f_sub, distance):
    f_sub[root][0] = 1  # co duy nhat chinh root co kc toi no la 0
    parent[root] = p
    for v in Adj[root]:
        if v not in parent:
            cal_f_sub(v, root, parent, Adj, f_sub, distance)
            for i in range(1, distance + 1):
                f_sub[root][i] += f_sub[v][i - 1]
    # print("Done visited ",root)
    topo.append(root)


def solver():
    import time
    start = time.time()
    # n, dist = map(int, input().split())
    n = 50000
    dist = 500
    Adj = {}
    for i in range(n): Adj[i] = []
    for i in range(n - 1):
        # a, b = map(lambda x: int(x) - 1, input().split())
        Adj[i].append(i + 1)
        Adj[i + 1].append(i)
    root = 0
    f_sub = [[0 for i in range(dist + 1)] for j in range(n)]  # 0 1 2 3 .... k
    parent = {}
    # print("thinh")
    cal_f_sub(root, None, parent, Adj, f_sub, dist)
    # print(f_sub)
    topo.reverse()
    # print(topo)
    f = [[0 for i in range(dist + 1)] for j in range(n)]
    f[root][0] = 1
    for i in range(dist + 1):
        f[root][i] = f_sub[root][i]
    for i in range(n):
        f[i][0] = 1
    for idx in range(1, len(topo)):
        u = topo[idx]
        for i in range(1, dist + 1):
            if i >= 2:
                f[u][i] = f_sub[u][i] + (f[parent[u]][i - 1] - f_sub[u][i - 2])
            else:
                f[u][i] = f_sub[u][i] + (f[parent[u]][i - 1])
    end = time.time()

    ans = 0
    for i in range(n):
        ans += f[i][dist]
    print(ans // 2)
    print(start - end)


solver()

File: Orange/W4_DevideAndConquer/7_Distance in Tree.cpp
#include<bits/stdc++.h>
using namespace std;
const int nax = 5e4 + 10;
int k;
long long int ans = 0;
vector<int> g[nax];
int visited[nax];
vector<int> topo;
long long int f[nax][510];
long long int f_sub[nax][510];
#define NOT_VISITED -100
void dfs(int v,int parent)
{
	f_sub[v][0] = 1;
	visited[v] = parent;
	for (auto u : g[v])
	{
		if (visited[u]!= NOT_VISITED) continue;
		dfs(u,v);
		for (int i = 1; i <= k; i++)
		{
			f_sub[v][i] += f_sub[u][i - 1];
		}
	}
	topo.push_back(v);
}


int main()
{

	//memset(visited, 0, sizeof(visited));
	for (size_t i = 0; i < sizeof(visited)/sizeof(visited[0]); i++)
	{
		visited[i] = NOT_VISITED;
	}
	memset(f_sub, 0, sizeof(f_sub));
	memset(f, 0, sizeof(f));
	for (int i = 0; i < nax; i++)
		g[i].clear();
	int n;
	cin >> n >> k;
	int u,v;
	for (int i = 0; i < n - 1; i++)
	{
		cin >> u >> v;
		u--;
		v--;
		{
			g[u].push_back(v);
			g[v].push_back(u);
		}

	}


	//int ans=0;
	int root = 0;
	dfs(root,-1);
	for (size_t i = 0; i <= k; i++)
	{
		f[0][i] = f_sub[0][i];

	}
	for (size_t i = 0; i < n; i++)
	{
		f[i][0] = 1;
	}
	reverse(topo.begin(), topo.end());
	for (size_t i = 1; i < topo.size(); i++)
	{
		u = topo.at(i);
		for (size_t j = 1; j <= k; j++)
		{
			if (j >= 2)
				f[u][j] = f_sub[u][j] + (f[visited[u]][j - 1] - f_sub[u][j - 2]);
			else
				f[u][j] = f_sub[u][j] + (f[visited[u]][j - 1]);
		}
	}
	for (size_t i = 0; i < n; i++)
	{
		ans += f[i][k];
	}
	cout << ans / 2 << endl;

}